### PR TITLE
[NETBEANS-96] New PAC Script evaluation environment

### PR DESCRIPTION
**Pull request by lbruun**
Implements a completely new PAC script environment to rectify the issues described in NETBEANS-96.

Also, Architecture documentation has been created for the module. (didn't exist previously).  There are a lot more test cases now. Previously there was virtually no testing of the PAC execution.

For now, the module requires Friend for everything. So even though it has been documented how one can plug in own PAC script evaluation environment, it is at this stage deliberately kept in Friend mode.

There are still many more things I have in the pipeline both for `ProxySelector` and for `Authenticator`. But I thought I would do it in incremental steps. Otherwise it gets too big for anyone to fully understand and thus review. Therefore this PR is solely about replacing the PAC script evaluation environment.

The new code requires Java 8. I get a few warnings on build, mainly related to my use of `@NonNull` annotation, but it doesn't seem to create problems. Happy to change.

### CODE REVIEW

File: core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+            
+            return new PacScriptEngine(engine, jsMainFunction);
+        } catch (ScriptException ex) {
+            throw new  PacParsingException(ex);
+        }
+    }
+    
+    private boolean getNashornJava8u40Available() {
+        try {
+            Class<?> klass = Class.forName("jdk.nashorn.api.scripting.NashornScriptEngineFactory");
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java]: Yes, it is a public API. The Nashorn which was in JDK8 prior to u40 was pretty useless. So we need a way to test if we have the newer Nashorn available (JDK8u40 onwards) or the older Nashorn (pre JDK8u40).

File: core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Stub for PAC Helpers. 
+ * 
+ * <p>
+ * The 'Helpers' are utility functions that the PAC script can make use of.
+ * 
+ * @author lbruun
+ */
+public abstract class PacHelperMethods implements PacHelperMethodsNetscape, PacHelperMethodsMicrosoft {
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java]: Yes. But also allows an implementation to choose to only implement the Netscape stuff. However my actual implementation implements both of them.

File: core.network/nbproject/project.properties [hunk 3]
```diff
@@ -16,7 +16,9 @@
 # under the License.
 
 is.autoload=true
-javac.source=1.6
+javac.source=1.8
 javac.compilerargs=-Xlint -Xlint:-serial
+javadoc.arch=${basedir}/arch.xml
+javadoc.apichanges=${basedir}/apichanges.xml
 release.external/nsProxyAutoConfig-mozilla2.0.js=modules/ext/nsProxyAutoConfig.js
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/nbproject/project.properties]: Correct. Will fix.

File: core.network/nbproject/project.xml [hunk 4]
```diff
@@ -1,169 +1,194 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/nbproject/project.xml]: Grr. Don't know how this happened. Will look into that.

File: nbbuild/build.properties [hunk 5]
```diff
@@ -188,6 +188,7 @@ config.javadoc.devel=\
 config.javadoc.friend=\
     o.n.bootstrap,\
     o.n.core,\
+	core.network,\
```
[the remark this answers is not in the document]
Comment (lbruun) [reply nbbuild/build.properties]: Yes.

**Comment by lbruun** [comment]
Matthias, your comments are a pleasure to work with. !! 
Amazing. Thanks. I'm on it.

File: core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.core.network.proxy.pac.impl;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java]: Will fix

File: core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public void clear() {
+        map.clear();
+    }
+    
+    private K findEvictionCandidate() {
+        // Finds minimum of all timestamps of all elements in the cache.
+        return map.entrySet().stream()
+                .min(Comparator.comparingLong( e -> e.getValue().lastUsed.get()))
+                .get().getKey();
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: Why wouldn't it be thread-safe?   map is `ConcurrentHashMap`. I'm confident it is "safe" (as in not throwing errors), but I guess your point is that the stream can (potentially) return a key which has been deleted in the meantime by another thread (two threads both calling `put` at the same time). But this will just mean that one of the threads will call `map.remove()` with a key that no longer exists and that's ok. (it is not an error). I think it is ok that the eviction may theoretically fail to delete something. It will fix itself on the next put operation.

File: core.network/src/org/netbeans/core/network/proxy/pac/datetime/DateRange.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+ * - day of month is 1-31. The range for this value is deliberately not
+ *    validated against date, meaning 31st of February is valid. This
+ *    creates a lenient evaluation.
+ * 
+ * Handles correctly the situation where a date range passes into next period.
+ * Examples:
+ * 
+ *   - Only month range is defined. monthMin=11 monthMax=02.  The period from 
+ *     December 1st to end of February.
+ *   - Only day of month range is defined. dateMin=27 dateMin=5. The days from the
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/datetime/DateRange.java]: Correct. Fixed.

File: core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (svatoun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: concurrentMap.entrySet() is not a copy, it is still backed by the map. So any parallel remove from the origin (concurrent)Map will be reflected in the Set, therefore invalidating Iterators and Streams derived from the entry set. The stream will eventually give a ConcurrentModificationException.

File: core.network/src/org/netbeans/core/network/proxy/pac/PacUtils.java [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        if (ipPrefix.indexOf('/') == -1) {
+            return false;
+        }
+        String[] parts = ipPrefix.trim().split("\\/");
+
+        int bitField = 0;
+        try {
+            bitField = Integer.parseInt(parts[1].trim());
+            // bitField < 8 doesn't make much sense in my book
+            if (!(bitField >= 8 && bitField <= 128)) {
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/PacUtils.java]: Well, values below 8 become a bit academic, but you are are right. There's no technical problem in allowing it, so will do.

**Comment by lbruun** [comment]
Unit tests fixed. 

- The `timeRange()` test failed because it did `assertTrue(timeRange(0, 23))`, and since you did your test at 23:17 it would fail the test. The test case was obviously wrong to begin with and his been fixed.

- The `dnsResolveEx()` test failed because it made assumptions about IPv6 being available on the test host and was in any case incorrect to begin with. All tests that use name lookup have now been changed to use the `FakeDns` class, so that tests become independent of the environment of test host. There was quite a few other tests that made such stupid assumptions.

Thx.

File: core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: @svatoun. Yes, I understand it is still backed by the map. However, the Javadoc for [entrySet](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#entrySet--) says:

> The view's spliterator reports Spliterator.CONCURRENT, Spliterator.DISTINCT, and Spliterator.NONNULL.

(the spliterator is used to create the stream, no?)

Note that this is specific to `ConcurrentHashMap`.  A standard `HashMap` doesn't have these attributes. 

Anyway, the way I understand [Spliterator.CONCURRENT](https://docs.oracle.com/javase/8/docs/api/java/util/Spliterator.html#CONCURRENT) is that is means we'll never get `ConcurrentModificationException`.

I'm far from certain here, so I´m listening!

[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: We can add `synchronized` to the `put()` method, even I think it is not needed.  This will for sure be acceptable from a performance point of view. The cache is expected to be used in situations where the there are many, many more reads than writes.

File: core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+     * a production PAC script. It is only intended for debugging and
+     * unit test purpose.
+     *
+     * <p>
+     * The default implementation simply writes to stderr.
+     *
+     * @param message text to log 
+     */
+    public void alert(String message) {
+        System.err.println(message);
```
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java]: The `PacHelperMethods`  is abstract. The actual implementation, `NbPacHelperMethods`,  now overrides this method and logs to logger with INFO level.

File: core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthiasblaesing) [reply core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java]: The second part of your reasoning won't work. To replace the `NbPackHelperMethod` you need to derive from `PacHelperMethods` as that abstract class declares, that it implements both `PacHelperMethodsNetscape` and `PacHelperMethodsMicrosoft`, but holds no implementations for their methods, so it falls to the implementor, who is now responsible for implementing both.

I don't say, that this is bad, just that the reasoning won't work :-)

File: core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (matthiasblaesing) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: Your argument about `Spliterator.CONCURRENT` is in line with the documentation of the ConcurrentHashMap:

> Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash  table at some point at or since the creation of the iterator/enumeration. They do not throw  ConcurrentModificationException.

so I would remove the synchronized from `evictIfFull`

[the remark this answers is not in the document]
Comment (svatoun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: I stand corrected: I didn't consider the origin of the Spliterator. Sorry for the confusion.

File: core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+    private PacJsEntryFunction testScriptEngineGeneric(String pacSource) throws PacParsingException {
+        Pattern pattern1 = Pattern.compile("function\\s+" + PacJsEntryFunction.STANDARD.getJsFunctionName());
+        Pattern pattern2 = Pattern.compile("function\\s+" + PacJsEntryFunction.IPV6_AWARE.getJsFunctionName());
+        if (!pattern2.matcher(pacSource).matches()) {
+            return PacJsEntryFunction.IPV6_AWARE;
+        }
+        if (!pattern1.matcher(pacSource).matches()) {
+            return PacJsEntryFunction.STANDARD;
+        }
+        throw new PacParsingException("Function " + PacJsEntryFunction.STANDARD.getJsFunctionName() + " or " + PacJsEntryFunction.IPV6_AWARE.getJsFunctionName() + " not found in PAC Script.");
```
Comment (matthiasblaesing) [inline core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java]: I walked through the code (at least partitally) and here was stopped. I checked rhino and this works ok (`ScriptEngine#get` return non-null):

```java
    public static void main(String[] args) throws Exception {
        ScriptEngineManager m = new ScriptEngineManager();
        ScriptEngine engine = m.getEngineByName("rhino");

        engine.eval("function halloWelt() {\nprint(\"Hallo\");\n}");
        
        System.out.println(engine.get("halloWelt"));
        
        engine.eval("halloWelt()");
    }
```

I think it is a reasonable expectation, that a JS ScriptEngine makes all variables declared in global scope available in the `ScriptContext.ENGINE_SCOPE`. So I basicly suggest to use the `isJsFunctionAvailable` also for the generic case and just skip the `ScriptObjectMirror` check. If someone is stupid enough to declare a variable with the name `FindProxyForURL` or `FindProxyForURLEx` in a PAC script, that is not a function he/she will do other mightily stupid things.

File: core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/PacHelperMethods.java]: Yep. You are right. I guess this changed somewhat during development. The _original_ idea was as described. I think there's still a logical benefit in keeping Netscape's original spec separated from Microsoft's later additions.

File: core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java (hunk 11 above)
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/proxy/pac/impl/NbPacScriptEvaluator.java]: Valid point. I'll fix.

File: core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/SimpleObjCache.java]: There's actually still a point to having the `synchronized` there. It not because of thread-safety because the code - even without the `synchronized` is now free from concurrency exceptions, However, it protects against the situation when multiple threads simultaneously arrive at the conclusion that eviction is required, so they all perform an evict, thereby essentially evicting _more_ elements than is strictly needed.

File: core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+     * 
+     * <p>
+     * Windows API equivalent: {@code GetComputerName()} function from 
+     * {@code Kernel32} library.
+     * 
+     * @return computer name
+     * @throws NativeException if there was an error executing the
+     *    system call.
+     */
+    public static String getComputerName() throws NativeException {
```
Comment (matthiasblaesing) [inline core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java]: This function is not used and I would remove this and the logger above. When this is done, the dependency on JNA Platform should also be removed. I noticed this, because I thought about moving the JNA function definitions for the function into core.network. JNA Platform weights > 2 MB and if we can get away without it, I think this would be good.

As you don't need it, please remove.

File: core.network/src/org/netbeans/core/network/utils/IpAddressUtils.java [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+        if (ipAddressStr == null || ipAddressStr.isEmpty()) {
+            return false;
+        }
+        if (!(isAsciiDigit(ipAddressStr.charAt(0)))) {
+            return false;
+        }
+        int dotPos = ipAddressStr.indexOf('.');
+        if (dotPos > 0) {
+            for(int i = 0; i < dotPos; i++) {
+                if (!(isAsciiDigit(ipAddressStr.charAt(0)))) {
```
Comment (matthiasblaesing) [inline core.network/src/org/netbeans/core/network/utils/IpAddressUtils.java]: The index should be _i_ not _0_. In general: is this really faster, than `IPV4_PATTERN.matcher(ipAddressStr).matches()`?

File: core.network/src/org/netbeans/core/network/utils/HostnameUtils.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+     *       although it is more than likely that Windows has cached this result 
+     *       on computer startup in its DNS Client Cache and therefore the 
+     *       result will be returned very fast.</li>
+     * </ul>
+     * 
+     * @return host name
+     * @throws NativeException if there was an error executing the
+     *    system call.
+     */
+    public static String getNetworkHostname() throws NativeException {
```
Comment (matthiasblaesing) [inline core.network/src/org/netbeans/core/network/utils/HostnameUtils.java]: I read through the hname packages and my reading: Everything quaks like Unix, just windows is different. So why not simplify this to:

```java
    public static String getNetworkHostname() throws NativeException {
        switch(Platform.getOSType()) {
            case Platform.WINDOWS:
                return org.netbeans.core.network.utils.hname.win.HostnameUtilsWin.getHostName(true);
            case Platform.MAC:
            case Platform.LINUX:
            case Platform.SOLARIS:
            default:
                return org.netbeans.core.network.utils.hname.unix.HostnameUtilsUnix.cLibGetHostname();
        }
    }
```

And remove the delegates in the platform packages.

File: core.network/src/org/netbeans/core/network/utils/IpAddressUtils.java (hunk 13 above)
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/IpAddressUtils.java]: The bug fixed. Thx.

Re use of regexp. The IPV4_PATTERN is used to check for full validity. And it requires a full IPv4 pattern, e.g. `#.#.#.#`  as opposed to say `#.#` which in Java world is also a valid IPv4 address. 

The two `looksLike`  methods have a very narrow use case. I've tried to explain that in the Javadoc (improved). They are not validity tests. If the user actually wants full validity check he should use `isValidIpv4Address()` as opposed to `looksLikeIpv4Literal()`. The former is actually now unused in the module but I've kept it for completeness and because it may have value for those implementing their own version of  PAC evaluator. This is also the reason why the two `looksLike` methods are not private, even if they have a very narrow use case.

Finding out if a string is a host name or an IP literal is somewhat difficult. The problem is that an all digit string is actually a valid hostname (what wise guy came up with that RFC??), but a digit is not allowed as the first character in a domain name. Example: "123.45" is a valid IPv4 address, but it is not a valid hostname, because of the '4'. However, the corner case value of "123" is indeterminate. This is explained in the Javadoc.

File: core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java (hunk 12 above)
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java]: True, that it is not used in *my* implementation. I used to use it, but then I switched to using `getHostName` instead.  There are pros and cons on these two Win32 methods and someone implementing differently than me may feel differently about it. So I kept it.

The dependency on JNA Platform will come back with a vengeance in the next iterations on the module where I improve further on the module. So no point in removing it now. Remember, as it says in the top text: The improved PAC script evaluator is just the beginning. To give you a taste: The next steps will for example add vastly improved config discovery and for this JNA Platform is needed.

File: core.network/src/org/netbeans/core/network/utils/HostnameUtils.java (hunk 14 above)
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/HostnameUtils.java]: Well, this is partly due to history. If I remember correctly, I used to use different functions on MacOSX and Linux. But through trials and tests I finally settled on using the same C lib function on everything but Windows. I may revert that.

The `hname` packages exist mainly as a way to get the host's _kernel name_, so that we can use that in an insane ride to find the host's own IP address as it is generally known on the network. This functionality is mandated for PAC evaluator. 

But when I started on the hname package, I thought of it as a general purpose package for everything related to the host's naming. For example, I wanted to add methods for getting the host's _display name_, e.g. "Joey's Laptop" and this is done very differently between OSes. This is another reason for the sub-packages per OS. I don't know if I'll ever get around to adding that. Right now the focus was on getting a working PAC evaluator as the existing one is broken. And frankly the host's display name has little usage except for use in About box and things like that.  :-)

I would like to keep the subpackages and the delegation for now even if it is a lot of ceremony for nothing.

File: core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java (hunk 12 above)
Comment (lbruun) [reply core.network/src/org/netbeans/core/network/utils/hname/win/HostnameUtilsWin.java]: About size: If that is a concern, how about the JNA project would consider splitting Platform jar into one jar per OS ?   :-)  COM in its own jar, of course, in order to further reduce size.

Anyways, I don't see a size of > 2MB. When I look in the `.nbm` file I see a pack200'ed jar of approx 178 KB. 

But if you feel it is important to get rid of JNA Platform lib then I'll be happy to extract the bits I need into the source. And I'll have to do the same in the next iterations of `core.network`, of course.

**Comment by matthiasblaesing** [comment]
I gave it another look I have one further comment on the code itself - the spec version of lib.jna.platform needs to be incremented, as the change in the friend-packages indicates that the package needs an update. Some of the files are also missing license headers. I made these here:

- based on a rebased+squashed changeset: 
https://github.com/matthiasblaesing/incubator-netbeans/tree/pr161-rebase
https://github.com/matthiasblaesing/incubator-netbeans/commit/6e21bcdd1c5d9f3eab4868b5ed1245c94b9c02be

- based on your branch:
https://github.com/matthiasblaesing/incubator-netbeans/tree/pr161-review
https://github.com/matthiasblaesing/incubator-netbeans/commit/5826080b5301846ff2e38cb2f004e0f470526cc4

One structural question: Is there a reason for the friend-exports? I don't see usage of the org.netbeans.core.network packages in org.netbeans.core. Is there a concrete use-case/usage you have planed? Is it really sensible to have a different implementation of the JS evaluation environment for the PAC, that is not better done in the core implementation?

From my perspective that would be the last two things - I noticed this while I prepared the merge.

**Comment by lbruun** [comment]
I've done the requested changes. (license headers, bumped libs.jna.platform version)

**Wrt frind-packages.**
It is correct that `org.netbeans.core` doesn't need to be friend on `org.netbeans.core.network`, but there needs to be at least one friend in `<friend-packages>`  for  the XML to be valid. If I leave out `<friend-packages>` completely then I loose two things:

- I do not get any Javadoc produced, thus any adopter will not know what is possible to do. 
- I have no way way to "document" which packages I intend (in a controller manner for now, hence friend restriction) to be exposed, and which I intend not be exposed.

In other words: The friend declaration is a trick. If there's another way to accomplish the same, then by all means.

The `<friend-packages>` section will become even more important in the next iterations, but even in next iteration I do not really have a specific friend that I can add, so `org.netbeans.core` kinda acts like a dummy in this respect.

**Comment by matthiasblaesing** [comment]
Thank you very much for the work you put in here and the time you took to answer my questions/comments.

I rebased the work onto master and squashed the changeset together to improve readability (from experience I know, that there is a need to read comments from commits in the past and a compact commit helps there).

A final thought about a public SPI (pluggable PAC evaluation):

- a separate SPI module should be created on which the
  `core.network` would depend on

- `PacJsEntryFunction` should be removed from 
  `org.netbeans.core.network.proxy.pac` package, as it is not
  used in the SPI

- `PacScriptEvaluatorNoProxy` and `PacUtils` should be removed from
  `org.netbeans.core.network.proxy.pac` package, as they are
  not part of the SPI, but utility classes

- Only the `org.netbeans.core.network.proxy.pac` package 
  should be exposed

- Usage of interface/class should be unified:

  `org.netbeans.core.network.proxy.pac.PacHelperMethods`
  `org.netbeans.core.network.proxy.pac.PacScriptEvaluator`
  `org.netbeans.core.network.proxy.pac.PacHelperMethodsMicrosoft`
  `org.netbeans.core.network.proxy.pac.PacHelperMethodsNetscape`
  `org.netbeans.core.network.proxy.pac.PacScriptEvaluatorFactory`

  should either be interfaces or abstract classes.
  
  With default methods interface can be evolved similar to  abstract classes.

**Comment by eirikbakke** [comment]
@lbruun Does this patch completely replace the old [NetBeansProxy2](https://bitbucket.org/phansson/netbeansproxy2) project?

(I've been including NetBeansProxy2 in my NetBeans Platform application for some years now, and was wondering if I can now drop it on the newer platform versions.)
